File: setup-db.js
#!/usr/bin/env node
/**
 * Database Setup Script — Creates SQLite database with seed data
 * Agent A03: SQLite Data Owner
 */

import Database from 'better-sqlite3';
import fs from 'fs';
import path from 'path';

const dbPath = 'src/data/locations.db';
const schemaPath = 'src/data/schema.sql';
const seedPath = 'src/data/seed-data.json';

console.log('🗄️  Setting up SQLite database...\n');

try {
  // Remove existing database if it exists
  if (fs.existsSync(dbPath)) {
    console.log(`Removing existing database: ${dbPath}`);
    fs.unlinkSync(dbPath);
  }

  // Create database connection
  const db = new Database(dbPath);
  console.log(`✅ Database created: ${dbPath}`);

  // Enable WAL mode for better concurrency
  db.pragma('journal_mode = WAL');

  // Read and execute schema
  const schema = fs.readFileSync(schemaPath, 'utf-8');
  const schemaStatements = schema
    .split(';')
    .map(s => s.trim())
    .filter(s => s.length > 0);

  console.log(`\nExecuting schema (${schemaStatements.length} statements)...`);
  for (const statement of schemaStatements) {
    db.exec(statement);
  }
  console.log('✅ Schema created successfully');

  // Read seed data
  const seedData = JSON.parse(fs.readFileSync(seedPath, 'utf-8'));
  console.log(`\nInserting seed data (${seedData.locations.length} locations)...`);

  // Prepare insert statement
  const insertStmt = db.prepare(`
    INSERT INTO locations (
      id, name, lat, lng, address, charger_type, connectors, power_kw, source, verified_date
    ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
  `);

  // Insert all locations
  const insertMany = db.transaction((locations) => {
    for (const loc of locations) {
      insertStmt.run(
        loc.id,
        loc.name,
        loc.lat,
        loc.lng,
        loc.address,
        loc.charger_type,
        loc.connectors,
        loc.power_kw,
        loc.source,
        loc.verified_date
      );
    }
  });

  insertMany(seedData.locations);
  console.log(`✅ Inserted ${seedData.locations.length} locations`);

  // Verify integrity
  console.log('\n🔍 Verifying database integrity...');
  const integrityResult = db.prepare('PRAGMA integrity_check').all();
  if (integrityResult[0]['integrity_check'] === 'ok') {
    console.log('✅ Database integrity: OK');
  } else {
    throw new Error('Database integrity check failed');
  }

  // Get row count
  const rowCount = db.prepare('SELECT COUNT(*) as count FROM locations').get();
  console.log(`✅ Locations table: ${rowCount.count} rows`);

  // Checkpoint WAL
  console.log('\n💾 Checkpointing WAL...');
  db.pragma('wal_checkpoint(TRUNCATE)');
  console.log('✅ WAL checkpointed successfully');

  // Close database
  db.close();
  console.log(`\n✅ Database setup complete: ${dbPath}`);
  console.log(`   Schema: src/data/schema.sql`);
  console.log(`   Seed data: src/data/seed-data.json`);
  console.log(`   File size: ${(fs.statSync(dbPath).size / 1024).toFixed(2)} KB`);

} catch (error) {
  console.error('❌ Database setup failed:');
  console.error(error.message);
  process.exit(1);
}

File: setup-db.py
#!/usr/bin/env python3
"""
Database Setup Script — Creates SQLite database with seed data
Agent A03: SQLite Data Owner
"""

import sqlite3
import json
import os

DB_PATH = 'src/data/locations.db'
SCHEMA_PATH = 'src/data/schema.sql'
SEED_PATH = 'src/data/seed-data.json'

print('🗄️  Setting up SQLite database...\n')

try:
    # Remove existing database if it exists
    if os.path.exists(DB_PATH):
        print(f'Removing existing database: {DB_PATH}')
        os.remove(DB_PATH)

    # Create database connection
    conn = sqlite3.connect(DB_PATH)
    cursor = conn.cursor()
    print(f'✅ Database created: {DB_PATH}')

    # Enable WAL mode for better concurrency
    cursor.execute('PRAGMA journal_mode = WAL')

    # Read and execute schema
    with open(SCHEMA_PATH, 'r') as f:
        schema = f.read()

    schema_statements = [s.strip() for s in schema.split(';') if s.strip()]
    print(f'\nExecuting schema ({len(schema_statements)} statements)...')

    for statement in schema_statements:
        cursor.execute(statement)

    print('✅ Schema created successfully')

    # Read seed data
    with open(SEED_PATH, 'r') as f:
        seed_data = json.load(f)

    locations = seed_data['locations']
    print(f'\nInserting seed data ({len(locations)} locations)...')

    # Insert all locations
    insert_sql = '''
        INSERT INTO locations (
            id, name, lat, lng, address, charger_type, connectors, power_kw, source, verified_date
        ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
    '''

    for loc in locations:
        cursor.execute(insert_sql, (
            loc['id'],
            loc['name'],
            loc['lat'],
            loc['lng'],
            loc['address'],
            loc['charger_type'],
            loc['connectors'],
            loc['power_kw'],
            loc['source'],
            loc['verified_date']
        ))

    print(f'✅ Inserted {len(locations)} locations')

    # Verify integrity
    print('\n🔍 Verifying database integrity...')
    integrity = cursor.execute('PRAGMA integrity_check').fetchone()[0]
    if integrity == 'ok':
        print('✅ Database integrity: OK')
    else:
        raise Exception(f'Database integrity check failed: {integrity}')

    # Get row count
    row_count = cursor.execute('SELECT COUNT(*) FROM locations').fetchone()[0]
    print(f'✅ Locations table: {row_count} rows')

    # Commit first
    conn.commit()

    # Checkpoint WAL
    print('\n💾 Checkpointing WAL...')
    cursor.execute('PRAGMA wal_checkpoint(TRUNCATE)')
    print('✅ WAL checkpointed successfully')

    # Close connection
    conn.close()

    # Get file size
    file_size_kb = os.path.getsize(DB_PATH) / 1024
    print(f'\n✅ Database setup complete: {DB_PATH}')
    print(f'   Schema: {SCHEMA_PATH}')
    print(f'   Seed data: {SEED_PATH}')
    print(f'   File size: {file_size_kb:.2f} KB')

except Exception as e:
    print(f'❌ Database setup failed:')
    print(f'   {str(e)}')
    exit(1)
